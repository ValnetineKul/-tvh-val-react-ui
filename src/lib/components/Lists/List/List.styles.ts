import { makeStyles } from '../../../../themes/core';

const useStyles = makeStyles<void, 'root'>({ name: 'List' })((theme, _params, classes) => ({
  root: {
    margin: 'unset',
    paddingLeft: 20,
    maxWidth: 530,
  },

  orderedList: {},

  unorderedList: {
    [`&.${classes.root}`]: {
      listStyle: 'none',
    },
  },
}));

export default useStyles;
