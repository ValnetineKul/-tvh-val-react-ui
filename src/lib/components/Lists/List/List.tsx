import type { FC } from 'react';
import React from 'react';

import Typography from '../../Typography';
import useStyles from './List.styles';

export interface ListProps {
  ordered?: boolean;
  className?: string;
}

const List: FC<ListProps> = ({ children, ordered = false, className }) => {
  const { classes, cx } = useStyles();
  const Tag = (ordered ? 'ol' : 'ul') as keyof JSX.IntrinsicElements;

  const rootClassName = cx(classes.root, className, {
    [classes.orderedList]: ordered,
    [classes.unorderedList]: !ordered,
  });

  const childrenWithProps = React.Children.map(children, (child) => {
    if (React.isValidElement(child)) return React.cloneElement<ListProps>(child, { ordered });
    return null;
  });

  return (
    <Typography component={Tag} className={rootClassName}>
      {childrenWithProps}
    </Typography>
  );
};

export default List;
