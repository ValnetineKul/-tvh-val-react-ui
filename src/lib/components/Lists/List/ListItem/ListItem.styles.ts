import { makeStyles } from '../../../../../themes/core';

const useStyles = makeStyles({ name: 'ListItem' })((theme) => ({
  root: {
    position: 'relative',
    marginBottom: theme.layoutSpacing['spacing/200'],

    '& > ul': {
      paddingLeft: theme.layoutSpacing['spacing/450'],
      marginTop: theme.layoutSpacing['spacing/200'],
    },
  },

  orderedRoot: {},

  unorderedRoot: {
    '&:before': {
      content: '""',
      display: 'inline-block',
      width: 4,
      height: 4,
      borderRadius: theme.borderRadius['borderRadius/circle'],
      backgroundColor: theme.color['icon/onSurfaceCurrent/action/default'],
      position: 'absolute',
      left: -11,
      top: 10,
    },
  },

  iconRoot: {
    display: 'flex',
    '& > div > ul': {
      paddingLeft: theme.layoutSpacing['spacing/000'],
      marginTop: theme.layoutSpacing['spacing/200'],
    },
  },

  icon: {
    color: theme.color['icon/onSurfaceCurrent/action/default'],
    marginRight: theme.layoutSpacing['spacing/300'],
  },
}));

export default useStyles;
