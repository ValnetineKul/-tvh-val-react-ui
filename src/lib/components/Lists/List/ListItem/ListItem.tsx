import type { FC, ReactNode } from 'react';
import React from 'react';

import useStyles from './ListItem.styles';
import Icon from '../../../Icon';

export interface ListItemProps {
  children: ReactNode;
  icon?: React.FunctionComponent<
    React.SVGProps<SVGSVGElement> & {
      title?: string;
    }
  >;
  isOrdered?: boolean;
  className?: string;
}

const ListItem: FC<ListItemProps> = ({ children, icon, isOrdered = false, className }) => {
  const { classes, cx } = useStyles();
  const listStyleClassName = isOrdered ? classes.orderedRoot : classes.unorderedRoot;
  const rootClassName = cx(classes.root, !icon ? listStyleClassName : classes.iconRoot, className);

  const renderChildren = () => {
    if (!icon) return children;

    return (
      <>
        {icon && <Icon icon={icon} className={classes.icon} />}
        <div>{children}</div>
      </>
    );
  };

  return <li className={rootClassName}>{renderChildren()}</li>;
};

export default ListItem;
