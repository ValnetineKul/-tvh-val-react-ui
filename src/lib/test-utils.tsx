import userEvent from '@testing-library/user-event';
import { fireEvent, render } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { theme } from '../themes/tvh';
import { createTheme, ThemeProvider, SurfaceProvider } from '../themes/core';

export const muiTheme = createTheme(theme);

const AllTheProviders: React.FC = ({ children }) => {
  return (
    <MemoryRouter>
      <ThemeProvider theme={muiTheme}>
        <SurfaceProvider color="100">{children}</SurfaceProvider>
      </ThemeProvider>
    </MemoryRouter>
  );
};

const customRender = (component: React.ReactElement, options?: Parameters<typeof render>[1]) =>
  render(component, { wrapper: AllTheProviders, ...options });

const customRenderHook: typeof renderHook = (hook, options) =>
  renderHook(hook, { wrapper: AllTheProviders, ...options });

export const initResponsiveTest = (size: number | 'Mobile' | 'Tablet' | 'Desktop' | 'LgDesktop' | 'XlDesktop') => {
  let width: number;
  if (size === 'Mobile') width = muiTheme.breakpoints.values.sm - 1;
  else if (size === 'Tablet') width = muiTheme.breakpoints.values.sm;
  else if (size === 'Desktop') width = muiTheme.breakpoints.values.md;
  else if (size === 'LgDesktop') width = muiTheme.breakpoints.values.lg;
  else if (size === 'XlDesktop') width = muiTheme.breakpoints.values.xl;
  else width = size;

  Object.defineProperty(window, 'innerWidth', {
    writable: true,
    configurable: true,
    value: width,
  });

  window.matchMedia = jest.fn().mockImplementation((query) => {
    const queryWidthInterval: string[] = query.match(/\d+(\.\d+)?/g);

    let matches;
    if (queryWidthInterval.length === 2) {
      matches = width >= +queryWidthInterval[0] && width < +queryWidthInterval[1];
    } else {
      const queryWidth = +queryWidthInterval[0];
      const isDown = /max/.test(query);

      matches = isDown ? width <= queryWidth : width >= queryWidth;
    }

    return {
      matches,
      media: query,
      onchange: null,
      addListener: jest.fn(),
      removeListener: jest.fn(),
    };
  });
};

const customUserEvent = Object.assign(userEvent, {
  mouseDown: fireEvent.mouseDown,
  mouseUp: fireEvent.mouseUp,
  mouseMove: fireEvent.mouseMove,
  scroll: fireEvent.scroll,
  error: fireEvent.error,
});

export { screen, within, waitFor, act, waitForElementToBeRemoved } from '@testing-library/react';
// override render method
export { customRender as render, customRenderHook as renderHook, customUserEvent as userEvent };
