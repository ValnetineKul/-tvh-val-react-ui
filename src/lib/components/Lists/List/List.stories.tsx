import type { ComponentProps } from 'react';
import React from 'react';
import type { StoryTemplate } from '../../../story-utils';
import { createMeta } from '../../../story-utils';
import { Check } from '../../Icon/icons/functional';
import List from './List';
import ListItem from './ListItem';

type Props = ComponentProps<typeof List>;
type Args = Omit<Props, 'children'>;

const Template = ((args: Args) => {
  return (
    <List {...args}>
      <ListItem>Unordered list item</ListItem>
      <ListItem>Unordered list item</ListItem>
      <ListItem>Unordered list item</ListItem>
      <ListItem>
        Unordered list item
        <List>
          <ListItem>Unordered list item</ListItem>
          <ListItem>Unordered list item</ListItem>
          <ListItem>Unordered list item</ListItem>
        </List>
      </ListItem>
    </List>
  );
}) as StoryTemplate<Args>;

export const Primitive = Template.bind({});
Primitive.args = {
  ordered: false,
};

Primitive.storyName = 'List';

const IconTemplate = ((args: Args) => {
  return (
    <List {...args}>
      <ListItem icon={Check}>Icon list item</ListItem>
      <ListItem icon={Check}>Icon list item</ListItem>
      <ListItem icon={Check}>Icon list item</ListItem>
      <ListItem icon={Check}>
        Icon list item
        <List>
          <ListItem icon={Check}>Icon list item</ListItem>
          <ListItem icon={Check}>Icon list item</ListItem>
          <ListItem icon={Check}>Icon list item</ListItem>
        </List>
      </ListItem>
    </List>
  );
}) as StoryTemplate<Args>;

export const IconPrimitive = IconTemplate.bind({});
IconPrimitive.args = {
  ...Primitive.args,
};

IconPrimitive.storyName = 'IconList';

export default createMeta({
  component: List,
  title: 'Components/Lists/List',
  argTypes: {
    ordered: {
      control: {
        // Remove this when ordered styles will be added
        disable: true,
      },
    },
  },
});
