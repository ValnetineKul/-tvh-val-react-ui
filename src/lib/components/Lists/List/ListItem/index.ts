import ListItem from './ListItem';

export type { ListItemProps } from './ListItem';
export default ListItem;
