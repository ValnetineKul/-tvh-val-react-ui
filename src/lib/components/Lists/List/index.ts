import List from './List';
import ListItem from './ListItem';

import type { ListProps } from './List';
import type { ListItemProps } from './ListItem';

export { ListItem };
export default List;
export type { ListProps, ListItemProps };
