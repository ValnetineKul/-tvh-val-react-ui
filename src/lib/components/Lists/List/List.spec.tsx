import React from 'react';
import List from './List';
import ListItem from './ListItem';
import { render } from '../../../test-utils';

describe('List', () => {
  it('Should render correctly', () => {
    const { container } = render(
      <List>
        <ListItem>List Item</ListItem>
      </List>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
