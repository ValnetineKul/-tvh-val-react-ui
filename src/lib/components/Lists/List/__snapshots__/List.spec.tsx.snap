// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List Should render correctly 1`] = `
<ul
  class="tss-0-Typography-weightRegular-ref tss-1-List-root-ref tss-15gxqo-Typography-root-Typography-weightRegular-List-root-List-unorderedList-Typography-body500"
>
  <li
    class="tss-u5ccdm-ListItem-root-ListItem-unorderedRoot"
  >
    List Item
  </li>
</ul>
`;
